src/rust/storage/seg/src/segments/segments.rs

```diff
@@ -111,15 +119,292 @@ impl Segments {
+            let fields_data = Box::new(pool.as_slice());
+
+            // create blank bytes to copy data into
+            let mut bytes = vec![0; fields_size];
+            // retrieve bytes from mmapped file
+            bytes.copy_from_slice(&fields_data[0..fields_size]);
+
+            // ----- Retrieve `headers` -----
+            let mut headers = Vec::with_capacity(0);
+            headers.reserve_exact(cfg_segments);
+
+            // retrieve each `SegmentHeader` from the raw bytes
+            for id in 0..cfg_segments {
+                let begin = header_size as usize * id;
+                let finish = begin + header_size as usize;
+
+                // cast bytes to `SegmentHeader`
+                let header = unsafe { *(bytes[begin..finish].as_mut_ptr() as *mut SegmentHeader) };
+                headers.push(header);
+            }
+
+            // ----- Retrieve `segment_size` -----
+            let mut offset = headers_size;
+            let mut end = offset + i32_size;
+
+            let segment_size = unsafe { *(bytes[offset..end].as_mut_ptr() as *mut i32) };
+            // TODO: compare `cfg_segment_size` and `segment_size`
+
+            // ----- Retrieve `free` -----
+            offset += i32_size;
+            end += u32_size;
+
+            let free = unsafe { *(bytes[offset..end].as_mut_ptr() as *mut u32) };
+
+            // ----- Retrieve `cap` -----
+            offset += u32_size;
+            end += u32_size;
+
+            let cap = unsafe { *(bytes[offset..end].as_mut_ptr() as *mut u32) };
+
+            // ----- Retrieve `free_q` -----
+            offset += u32_size;
+            end += free_q_size;
+
+            let free_q = unsafe { *(bytes[offset..end].as_mut_ptr() as *mut Option<NonZeroU32>) };
+
+            // ----- Retrieve `flush_at` -----
+            offset += free_q_size;
+            end += flush_at_size;
+
+            let flush_at = unsafe { *(bytes[offset..end].as_mut_ptr() as *mut Instant) };
+
+            // ----- Re-initialise `evict` -----
+
+            let evict_policy = builder.evict_policy;
+            let evict = Eviction::new(cfg_segments, evict_policy);
+
+            SEGMENT_CURRENT.set(cap as _);
+            SEGMENT_FREE.set(free as _);
+
+            Self {
+                headers: headers.into_boxed_slice(),
+                data,
+                segment_size,
+                free,
+                cap,
+                free_q,
+                flush_at,
+                evict: Box::new(evict),
+                data_file_backed: true,
+                fields_copied_back: true,
+            }
+        } else {
+            Segments::from_builder_new(builder)
+        }
+    }
+
+    /// Demolishes the segments by flushing the `Segments.data` to PMEM
+    /// (if filed backed) and storing the other `Segments` fields' to
+    /// PMEM (if a path is specified)
+    pub fn demolish(&self, segments_fields_path: Option<PathBuf>, heap_size: usize) -> bool {
```

Comment:
Same here, do we need a demolish or can this be `Drop` trait? Can we reduce the duplication between restore and demolish?